File: javascript/perfect-numbers/perfect-numbers.js
export const classify = (number) => {
    if (number <= 0) {
        throw new Error('Classification is only possible for natural numbers.');
    }

    let sum = aliquotSum(number);

    return sum === number ? 'perfect'
            : sum > number ? 'abundant' : 'deficient';
};

const aliquotSum = (number) => {
    let sum = 0;

    for (let i = 1; i <= number / 2; i++) {
        if (number % i == 0) {
            sum += i;
        }
    }

    return sum;
};

File: elixir/perfect-numbers/lib/perfect_numbers.ex
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, atom} | {:error, String.t()}
  def classify(number) when number < 1,
    do: {:error, "Classification is only possible for natural numbers."}

  def classify(number) do
    a_sum = aliquot_sum(number)

    cond do
      a_sum == number -> {:ok, :perfect}
      a_sum > number -> {:ok, :abundant}
      true -> {:ok, :deficient}
    end
  end

  defp aliquot_sum(1), do: 0

  defp aliquot_sum(number) do
    Enum.reduce(1..(number - 1), MapSet.new(), fn x, acc ->
      if rem(number, x) == 0 do
        MapSet.put(acc, x)
      else
        acc
      end
    end)
    |> Enum.sum()
  end
end
